_posts/2016-10-12-edsls-part-2.md

```diff
@@ -0,0 +1,413 @@
+---
+layout: post
+title: EDSLs as functions
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+*This is the second of a series of articles on "Monadic EDSLs in Scala."*
+
+Perhaps the most direct way to start writing an EDSL is to start writing
+functions. Let's say we want a language for talking about sets of integers.
+
+```scala
+trait SetLang {
+  def add(i: Int, set: Set[Int]): Set[Int]
+  def remove(i: Int, set: Set[Int]): Set[Int]
+  def exists(i: Int, set: Set[Int]): Boolean
+}
+```
+
+This works.. to the extent that we want only to work with
+`scala.collection.Set`s. As it stands we cannot talk about
+other implementations of sets, bloom filters, or sets controlled by
+remote programs. Our language isn't *abstract* enough, so let's remove
+all traces of `Set`.
+
+```scala
+trait SetLang[F[_]] {
+  def add(i: Int, set: F[Int]): F[Int]
+  def remove(i: Int, set: F[Int]): F[Int]
+  def exists(i: Int, set: F[Int]): Boolean
+
+  // Given unknown F we no longer know how to create an empty set
+  // so we add the capability to our language
+  def empty: F[Int]
+}
+```
+
+We've parameterized our language with a [higher-kinded type][hkt] which
+represents the context of our set. We can then write mini-programs which talks
+about sets without regards to `F`, which could be `scala.collection.Set`,
+`java.util.Set`, or `type RemoteSet[A] = Future[Set[A]]`.
+
+```scala
+def program[F[_]](lang: SetLang[F]): Boolean = {
+  import lang._
+  exists(10, remove(5, add(10, add(5, empty))))
+}
+```
+
+Interpretation of our language is straightforward, by implementing `SetLang` and
+passing in the corresponding implementation to `program`.
+
+Our language is still limited. `F` is arbitrarily designated to
```

Comment:
The problem is that the language is too powerful, not that it's limited. The capabilities it provides are so strong as to prevent you from writing the kinds of interpreters you are interested in writing, but there is no “better than” rule here—your redesign admits those interpreters only by forbidding many programs that formerly could be written in `SetLang`.